### PR TITLE
Port :sets.is_subset/2 to JS

### PR DESCRIPTION
**Pull request by NickBoyadjian**
## Summary by CodeRabbit

* **New Features**
  * Added subset comparison: new is_subset/2 returns true/false, treats an empty first set as a subset, and correctly handles equal and proper-subset cases.
* **Tests**
  * Added comprehensive tests for empty sets, equal and proper subsets, and invalid-input error cases to validate behavior and error handling.

<sub>✏️ Tip: You can customize this high-level summary in your review settings.</sub>

### CODE REVIEW

File: assets/js/erlang/sets.mjs [hunk 1]
```diff
@@ -86,6 +86,39 @@ const Erlang_Sets = {
...
+    if (!Type.isMap(set1)) {
+      Interpreter.raiseFunctionClauseError(
+        Interpreter.buildFunctionClauseErrorMsg(":sets.to_list/1", [set1]),
+      );
+    }
+    if (!Type.isMap(set2)) {
+      Interpreter.raiseFunctionClauseError(
+        Interpreter.buildFunctionClauseErrorMsg(":sets.to_list/1", [set2]),
+      );
+    }
```
[the remark this answers is not in the document]
Comment (NickBoyadjian) [reply assets/js/erlang/sets.mjs:143]: fixed

File: test/javascript/erlang/sets_test.mjs [hunk 2]
```diff
@@ -258,4 +258,102 @@ describe("Erlang_Sets", () => {
...
+        ":sets.to_list/1",
+        [atomAbc],
+      );
+
+      assertBoxedError(
+        () => is_subset(Erlang_Sets["new/1"](opts), atomAbc),
+        "FunctionClauseError",
+        expectedMessage,
+      );
+    });
```
[the remark this answers is not in the document]
Comment (NickBoyadjian) [reply test/javascript/erlang/sets_test.mjs:546]: fixed

File: assets/js/erlang/sets.mjs [hunk 3]
```diff
@@ -86,6 +86,45 @@ const Erlang_Sets = {
...
+      return subset.every((subItem) =>
+        superset.some(
+          (superItem) => Interpreter.compareTerms(subItem, superItem) == 0,
+        ),
+      );
+    };
+
+    const set1Array = Erlang_Sets["to_list/1"](set1).data;
+
+    if (set1Array.length == 0) {
```
Comment (bartblast) [inline assets/js/erlang/sets.mjs]: This should go before the actual check (const isSubset...) and changed to:
`if (set1.data.length === 0)`

File: assets/js/erlang/sets.mjs [hunk 4]
```diff
@@ -86,6 +86,45 @@ const Erlang_Sets = {
...
+    if (!Type.isMap(set2)) {
+      Interpreter.raiseFunctionClauseError(
+        Interpreter.buildFunctionClauseErrorMsg(":sets.is_subset/2", [
+          set1,
+          set2,
+        ]),
+      );
+    }
+
+    const isSubset = (subset, superset) => {
```
Comment (bartblast) [inline assets/js/erlang/sets.mjs]:
```js
return Type.boolean(sets.data.every((item) => Erlang_Sets["is_element/2"](item, set2).value))
```

(and add `:sets.is_element/2` to Deps both here and in call_graph.ex)

File: assets/js/erlang/sets.mjs [hunk 5]
```diff
@@ -86,6 +86,45 @@ const Erlang_Sets = {
   },
   // End to_list/1
   // Deps: [:maps.keys/1]
+
+  // Start is_subset/2
+  "is_subset/2": (set1, set2) => {
+    if (!Type.isMap(set1)) {
```
Comment (bartblast) [inline assets/js/erlang/sets.mjs]: Both error messages are the same so this could be:
`if (!Type.isMap(set1) || !Type.isMap(set2))`

File: test/javascript/erlang/sets_test.mjs [hunk 6]
```diff
@@ -258,4 +258,102 @@ describe("Erlang_Sets", () => {
       );
     });
   });
+
+  describe("is_subset/2", () => {
+    const is_subset = Erlang_Sets["is_subset/2"];
+
+    it("should always return true if set1 is an empty set", () => {
+      const set1 = Erlang_Sets["new/1"](opts);
+      const set2 = Erlang_Sets["new/1"](opts);
```
Comment (bartblast) [inline test/javascript/erlang/sets_test.mjs:474]: Please add another test where set1 is empty and set2 is non-empty.

File: test/javascript/erlang/sets_test.mjs [hunk 7]
```diff
@@ -258,4 +258,102 @@ describe("Erlang_Sets", () => {
...
+      const result = is_subset(set1, set2);
+
+      assert.deepStrictEqual(result, Type.boolean(true));
```
Comment (bartblast) [inline test/javascript/erlang/sets_test.mjs]: `assetBoxedTrue` here and in other tests

File: test/javascript/erlang/sets_test.mjs [hunk 8]
```diff
@@ -258,4 +258,102 @@ describe("Erlang_Sets", () => {
...
+      const set2 = Erlang_Sets["from_list/2"](
+        Type.list([Type.integer(1), Type.integer(2)]),
+        opts,
+      );
+      const result = is_subset(set1, set2);
+
+      assert.deepStrictEqual(result, Type.boolean(true));
+    });
+
+    it("should work with sets of tuples", () => {
```
Comment (bartblast) [inline test/javascript/erlang/sets_test.mjs]: this is redundant

File: assets/js/erlang/sets.mjs (hunk 4 above)
Comment (NickBoyadjian) [reply assets/js/erlang/sets.mjs]: Sorry if I'm missing something but isn't `Erlang_Sets["is_element/2"]` not implemented yet?

File: assets/js/erlang/sets.mjs (hunk 5 above)
Comment (NickBoyadjian) [reply assets/js/erlang/sets.mjs]: :heavy_check_mark:

File: assets/js/erlang/sets.mjs (hunk 3 above)
Comment (NickBoyadjian) [reply assets/js/erlang/sets.mjs]: :heavy_check_mark:

File: test/javascript/erlang/sets_test.mjs (hunk 6 above)
Comment (NickBoyadjian) [reply test/javascript/erlang/sets_test.mjs:474]: :heavy_check_mark:

File: test/javascript/erlang/sets_test.mjs (hunk 7 above)
Comment (NickBoyadjian) [reply test/javascript/erlang/sets_test.mjs]: :heavy_check_mark:

File: test/javascript/erlang/sets_test.mjs (hunk 8 above)
Comment (NickBoyadjian) [reply test/javascript/erlang/sets_test.mjs]: do you want me to just remove this test?

Comment (bartblast) [reply test/javascript/erlang/sets_test.mjs]: Yes, please.

**Comment by bartblast** [comment]
Also, please pull new changes from dev, @NickBoyadjian.

File: test/javascript/erlang/sets_test.mjs (hunk 8 above)
Comment (NickBoyadjian) [reply test/javascript/erlang/sets_test.mjs]: :heavy_check_mark:

File: assets/js/erlang/sets.mjs (hunk 4 above)
Comment (NickBoyadjian) [reply assets/js/erlang/sets.mjs]: that didn't quite work like that but I got it to work with the `is_element/2` function. Let me know if you see any issues with how I implemented it!

File: test/javascript/erlang/sets_test.mjs [hunk 9]
```diff
@@ -302,4 +302,88 @@ describe("Erlang_Sets", () => {
...
+      );
+
+      assertBoxedError(
+        () => is_subset(atomAbc, set123),
+        "FunctionClauseError",
+        expectedMessage,
+      );
+    });
+
+    it("raises FunctionClauseError if the second argument is not a set", () => {
```
Comment (bartblast) [inline test/javascript/erlang/sets_test.mjs:537]: This test doesn't match the Elixir test (the goal is to have JS <> Elixir consistency).

File: test/elixir/hologram/ex_js_consistency/erlang/sets_test.exs [hunk 10]
```diff
@@ -166,4 +166,77 @@ defmodule Hologram.ExJsConsistency.Erlang.SetsTest do
...
+
+  describe "is_subset/2" do
+    setup do
+      empty_set = :sets.new(version: 2)
+      set123 = :sets.from_list([1, 2, 3], version: 2)
+
+      [empty_set: empty_set, set123: set123]
+    end
+
+    test "should always return true if set1 is an empty set", ctx do
```
Comment (bartblast) [inline test/elixir/hologram/ex_js_consistency/erlang/sets_test.exs]: Could you not use the "should (...)" prefix? Here "returns true (...)" is ok, etc.

Comment (bartblast) [inline test/elixir/hologram/ex_js_consistency/erlang/sets_test.exs]: Instead of ctx, please use match, e.g.
`test "...", %{empty_set: empty_set, set_123: set_123}`
etc.

File: test/elixir/hologram/ex_js_consistency/erlang/sets_test.exs [hunk 11]
```diff
@@ -166,4 +166,77 @@ defmodule Hologram.ExJsConsistency.Erlang.SetsTest do
       end
     end
   end
+
+  describe "is_subset/2" do
+    setup do
+      empty_set = :sets.new(version: 2)
+      set123 = :sets.from_list([1, 2, 3], version: 2)
+
+      [empty_set: empty_set, set123: set123]
```
Comment (bartblast) [inline test/elixir/hologram/ex_js_consistency/erlang/sets_test.exs]: For indexes please use underscore:
set123 -> set_123

File: test/elixir/hologram/ex_js_consistency/erlang/sets_test.exs [hunk 12]
```diff
@@ -166,4 +166,77 @@ defmodule Hologram.ExJsConsistency.Erlang.SetsTest do
...
+
+    test "should return false if not all elements in set1 are in set2", ctx do
+      set1 = :sets.from_list([1], version: 2)
+      refute :sets.is_subset(set1, ctx.empty_set)
+    end
+
+    test "should return true if both sets are the same" do
+      set1 = :sets.from_list([1, 2], version: 2)
+      set2 = :sets.from_list([1, 2], version: 2)
+      assert :sets.is_subset(set1, set2)
```
Comment (bartblast) [inline test/elixir/hologram/ex_js_consistency/erlang/sets_test.exs]: Since we're testing exact result, could you test actual return value, e.g.
`assert :sets.is_subset(set1, set2) == true`
(assert will accept any truthy value)

File: test/javascript/erlang/sets_test.mjs [hunk 13]
```diff
@@ -302,4 +302,88 @@ describe("Erlang_Sets", () => {
...
+      const set2 = Erlang_Sets["from_list/2"](
+        Type.list([Type.integer(1), Type.integer(2)]),
+        opts,
+      );
+      const result = is_subset(set1, set2);
+
+      assertBoxedTrue(result);
+    });
+
+    it("raises FunctionClauseError if the first argument is not a set", () => {
```
Comment (bartblast) [inline test/javascript/erlang/sets_test.mjs:526]: This test doesn't match the Elixir test (the goal is to have JS <> Elixir consistency).

**Review by bartblast — CHANGES_REQUESTED** [review]

File: test/javascript/erlang/sets_test.mjs (hunk 13 above)
Comment (NickBoyadjian) [reply test/javascript/erlang/sets_test.mjs:526]: I will keep looking but figured I'd ask because I'm racking my head a little here. This is what the error in elixir looks like when calling `:sets.is_subset/2`
```
iex(1)> :sets.is_subset(1, 2)
** (FunctionClauseError) no function clause matching in :sets.fold/3

    The following arguments were given to :sets.fold/3:

        # 1
        #Function<9.49235441/2 in :sets.is_subset/2>

        # 2
        true

        # 3
        1

    (stdlib 6.2) sets.erl:588: :sets.fold/3
    iex:1: (file)
```

I'm not entirely sure how to capture that first argument  (the Function part). I guess erlang under the hood is using `:sets.fold/3` to get the result. Have you run into anything similar in this project?
@bartblast

Comment (NickBoyadjian) [reply test/javascript/erlang/sets_test.mjs:526]: I figure using `sets:fold` in the javascript implementation would probably clean that up :thinking:

Comment (bartblast) [reply test/javascript/erlang/sets_test.mjs:526]: @NickBoyadjian Yes, please use `:sets.fold/3` in the error message, but without the args details. This will do:
```js
Interpreter.buildFunctionClauseErrorMsg(":sets.fold/3")
```
(we will not display these args dump in the client-side version)

Comment (NickBoyadjian) [reply test/javascript/erlang/sets_test.mjs:526]: Would it make sense for me to wait for the `sets.fold` PR to go through, and then use that in my implementation? That's what my plan was

Comment (bartblast) [reply test/javascript/erlang/sets_test.mjs:526]: How you plan to use `:fold/3` here?

Comment (NickBoyadjian) [reply test/javascript/erlang/sets_test.mjs:526]: Sorry, not in the test but the actual implementation of the `is_subset` function.

Comment (bartblast) [reply test/javascript/erlang/sets_test.mjs:526]: Yes, how you plan to use it?

Comment (NickBoyadjian) [reply test/javascript/erlang/sets_test.mjs:526]: In elixir it would look like this
```elixir
:sets.fold(fn x, _result -> :sets.is_element(x, set2) end, true, set1)
```

Comment (bartblast) [reply test/javascript/erlang/sets_test.mjs:526]: A more optimal JS implementation would return early if any element was not part of set2.

Comment (NickBoyadjian) [reply test/javascript/erlang/sets_test.mjs:526]: Agreed. I'm not sure how similar you want to keep things to their erlang implementations. Let me know if you prefer me to use fold or just kind of fake the error message

Comment (bartblast) [reply test/javascript/erlang/sets_test.mjs:526]: We're after consistent behaviour not consistent implementation. Take at look at this: https://github.com/bartblast/hologram/issues/532#issuecomment-3725227738

Comment (NickBoyadjian) [reply test/javascript/erlang/sets_test.mjs:526]: Thanks for the explanation. I think that's a good pragmatic approach. I'll hopefully have some time this weekend to wrap this up. I'll ping you when it's ready for re-review!

Comment (NickBoyadjian) [reply test/javascript/erlang/sets_test.mjs:526]: @bartblast This is ready for re-review now

File: assets/js/erlang/sets.mjs [hunk 14]
```diff
@@ -129,6 +129,33 @@ const Erlang_Sets = {
...
+    return Type.boolean(
+      set1Array.every(
+        (item) =>
+          Interpreter.compareTerms(
+            Erlang_Sets["is_element/2"](item, set2),
+            Type.boolean(true),
+          ) == 0,
+      ),
+    );
+  },
```
[the remark this answers is not in the document]
Comment (bartblast) [reply assets/js/erlang/sets.mjs:156]: We can ignore this one.

**Comment by bartblast** [comment]
@NickBoyadjian CI reports formatting issue, could you run: `mix f`?

**Comment by NickBoyadjian** [comment]
@bartblast all checks are passing now

**Comment by bartblast** [comment]
Thank you @NickBoyadjian! 💜
